src/doc/src/reference/build-scripts.md

```diff
@@ -250,6 +254,45 @@ identifier, the value should be a string.
 [conditional compilation]: ../../reference/conditional-compilation.md
 [option-cfg]: ../../rustc/command-line-arguments.md#option-cfg
 
+### `cargo::rustc-check-cfg=CHECK_CFG` {#rustc-check-cfg}
+
+The `rustc-check-cfg` instruction tells Cargo to pass the given expected
+configs to the [`--check-cfg` flag][option-check-cfg] to the compiler. This is
+used for compile-time detection of unexpected [conditional compilation] configs.
```

Comment:
Unsure about the other ones but this feels backwards, starting with / focusing on the implementation (which people likely won't understand) and then saying why its useful.

Reply:
I can understand that it can feel backwards, personally it doesn't bother me.

Maybe it can be written like this; not sure if it's better.
```markdown
The `rustc-check-cfg` instruction is used for compile-time detection of unexpected
[conditional compilation] configs by telling Cargo to pass the given expected
configs to the [`--check-cfg` flag][option-check-cfg] to the compiler.
```

Reply:
People scan documentation.  The start of the description will either have people skip over it completely or they can misunderstand the entire behavior based off of that scan.

This flipped the statement but the statement can easily be misread.  This is saying this is for compile-time detection, which can read like this enables compile time detection.  This is for defining keys/values for conditional compilation configs so they don't get reported as unexpected.

Reply:
I reworked this part of the documention to include add an explaination of what it's doing and why it is doing so.

It looks like this:

-----

### `cargo::rustc-check-cfg=CHECK_CFG`

This instruction is related to checking the conditional configurations of the
crate at compile time. That is, checking that every _reachable_ `#[cfg]` matches
the expected config names and values.

This `rustc-check-cfg` instruction increase the list of expected config names
and values by telling Cargo which custom configs should be expected in addition
to Cargo features and [well known cfgs][well-known-cfgs].

It is backed by the [`--check-cfg` flag][option-check-cfg] of the compiler.

-----

I don't know if we should put more emphasis into why it's useful (not that is very explicit either).

Let me know what you think.

Reply:
@epage Are you satisfied with the above changes?

Reply:
Still feels like the lead-in gives the wrong impression about what this is about

How about

---

### `cargo::rustc-check-cfg=CHECK_CFG` {#rustc-check-cfg}

Add to the list of expected config names values that used when checking reachable `cfg` expressions.

For details on the syntax of `CHECK_CFG`, see rustc's [`--check-cfg` flag][option-check-cfg].  See also `<link to rustc lint>`.

The instruction can be used like this:

...

---

Reply:
Ed's suggestion is exactly what I proposed in <https://github.com/rust-lang/cargo/pull/13571#discussion_r1585451270> and <https://github.com/rust-lang/cargo/pull/13571#discussion_r1585459006>. Let me resolve those two then :)

Reply:
Ok, I took your suggestion; just tweaked it for my non native english eyes.

------

Add to the list of expected configs the names and values used when checking the _reachable_
cfg expressions.

For details on the syntax of `CHECK_CFG`, see `rustc` [`--check-cfg` flag][option-check-cfg].
See also the [`unexpected_cfgs`][unexpected-cfgs] lint.

Reply:
I think the change was

> Add to the list of expected **config names values that** used when checking reachable cfg expressions.

to

> Add to the list of expected **configs the names and values** used when checking the reachable
cfg expressions.

Re-arranging that to make clear the what the components are, you get "adding names and values to expected configs".  I find the concept of "expected configs" to be odd" and instead see this as us adding expected config names and values.

Of course mine had some grammatical mistakes.  Correcting for that, I propose:

> Add to the list of expected **config names and values that is** used when checking reachable cfg expressions.

Thoughts?

Reply:
While writing down my (overly long) response, I found myself switch back and forth between "this is fine" and I prefer "expected configs". But seeing that we already use "expected config names and values" is other places, I don't have an objection.

The only thing I still find a bit weird with your sentence is the *what?*, *what is being added to the list?*

Reply:
For me, that is implied with the directive itself, `CHECK_CFG`.  If you want to be more explicit, that is fine.